### ISSUE TITLE
Efficient aggregation for attention GNN

### ISSUE THREAD

**Issue by yhenon** [rusty1s/pytorch_geometric#1024]
## 🚀 Feature
I am attempting to write an implementation of the attention GNN used in:
https://arxiv.org/abs/1911.11763
https://arxiv.org/abs/2002.03983

(currently SOTA on homography estimation and point cloud matching, respectively)

## Motivation

The aggregation mechanism involves concatenating the messages, and taking the softmax across the messages. The default aggregation mechanisms won't support that, requiring a very slow for-loop

## Additional context

A simple implementation of single0headed attention as described in the papers above looks like this:

```
    def forward(self, x, edge_index, size=None):

        q = self.fc1(batch.x)
        k = self.fc2(batch.x)
        v = self.fc3(batch.x)

        propagated = self.propagate(batch.edge_index, size=None, x=batch.x, q=q, k=k, v=v, batch=batch.batch)
        return propagated

    def message(self, q, k, v, v_i, v_j, q_i, q_j, k_i, k_j, edge_index):
        alpha = (q_i * k_j).sum(dim=1) / 16
        return alpha, v_j

    def update(self, aggr_out):
        return aggr_out

    def aggregate(self, inputs, index, dim_size=None):

        output = []
        alpha = inputs[0]
        v_j = inputs[1]

        self.alpha = []

        for idx in range(index.max() + 1):
            indices = torch.where(index == idx)[0]
            alpha = torch.nn.functional.softmax(alpha[indices], dim=0)
            self.alpha.append(alpha)
            val = (alpha.unsqueeze(1) * v_j[indices, :]).sum(dim=0)
            output.append(val.unsqueeze(0))
        return torch.cat(output, dim=0)
```

Any suggestions on making this somewhat efficient?

**Comment by yhenon** [comment]
Just noticed the provided softmax operator. Closing this.
